### PR TITLE
[Extension Installer] Add implementation for extension usage detection from Siddhi app

### PR DESCRIPTION
**Pull request by senthuran16**
## Purpose
$subject, from `@source`, `@sink`, and `@store` annotations.

**WIP:**
> - Add implementation to install multiple extensions at once.

To support extension detection, the following _optional_ extension identifiers have been introduced to the extensions installer's configuration file.

> An extension for a particular usage in a Siddhi app, is matched in the same order of the below mentioned identifiers. When no match is found after checking by all of these identifiers, match by name is done (the existing implementation).

**1. Unique Attribute Value Identifier**
Extensions that share the common `type`, but identifiable based on values of unique attributes.
eg:
- `type='rdbms'` (common).
- `jdbc.driver.name='com.mysql.jdbc.Driver'` for MySQL.
- `jdbc.driver.name='org.postgresql.Driver'` for Postgres.

Sample configuration:
```
"rdbms-mysql": {
    "extension": {...},

    "identifier": {
      "type": "rdbms",
      "uniqueAttribute": "jdbc.driver.name",
      "uniqueAttributeValue": "com.mysql.jdbc.Driver"
    },

    "dependencies": [...]
}
```

**2. Unique Attribute Value Regex Identifier**
Extensions that share the common `type`, but identifiable based on the regex of values of unique attributes.
eg:
- `type='cdc'` (common).
- `url='jdbc:mysql://...'` for MySQL.
- `url='jdbc:postgresql://...'` for Postgres.

Sample configuration:
```
"cdc-mysql": {
    "extension": {...},

    "identifier": {
      "type": "cdc",
      "uniqueAttribute": "url",
      "uniqueAttributeValueRegex": "jdbc:mysql:(.+)"
    },

    "dependencies": [...]
}
```

**3. Alternative Type Identifier**
Extensions such that, same extension is denoted by different `type`s.
eg:
All of the following denote the `http` extension.
- `type='http'`
- `type='http-service-response'`
- `type='http-call-response'`

Sample configuration:
```
"http": {
    "extension": {...},

    "identifier": {
      "alternativeTypes": [
        "http-call",
        "http-request",
        "http-response",
        "http-service-response",
        "http-call-response",
        "http-request",
        "http-response",
        "http-service"
      ]
    },

    "dependencies": [...]
}
```

### CODE REVIEW

File: components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/config/mapping/models/ExtensionConfig.java [hunk 1]
```diff
@@ -28,28 +29,30 @@
...
+    private List<DependencyConfig> dependencies = new ArrayList<>();
 
     public List<DependencyConfig> getManuallyInstallableDependencies() {
-        if (dependencies != null) {
-            return dependencies.stream().filter(dependency -> !dependency.isAutoDownloadable())
-                .collect(Collectors.toList());
-        }
-        return Collections.emptyList();
+        return dependencies.stream()
+            .filter(dependency -> !dependency.isAutoDownloadable())
```
Comment (AnuGayan) [inline components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/config/mapping/models/ExtensionConfig.java]: can't we make it to two lines?

File: components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+        Optional<Map.Entry<String, ExtensionConfig>> matchedExtensionEntry =
+            extensionConfigs.entrySet().stream()
+                .filter(extension -> {
+                    ExtensionIdentifierConfig extensionIdentifier = extension.getValue().getIdentifier();
+                    if (extensionIdentifier != null && extensionIdentifier.isAlternativeTypeValid()) {
+                        return extensionIdentifier.getAlternativeTypes().contains(usage.getProperties().get(TYPE_KEY));
+                    }
+                    return false;
+                })
+                .findFirst();
```
Comment (AnuGayan) [inline components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java]: shall we move this to 181

File: components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+                        String uniqueAttribute = extensionIdentifier.getUniqueAttribute();
+                        String uniqueAttributeValueRegex = extensionIdentifier.getUniqueAttributeValueRegex();
+                        return (
+                            Objects.equals(usage.getProperties().get(TYPE_KEY), extensionIdentifier.getType()) &&
+                                usage.getProperties().get(uniqueAttribute) != null &&
+                                usage.getProperties().get(uniqueAttribute).matches(uniqueAttributeValueRegex));
+                    }
+                    return false;
+                })
+                .findFirst();
```
Comment (AnuGayan) [inline components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java]: shall we move this to 152

Comment (AnuGayan) [reply components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java]: Check other places as well

File: components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/config/mapping/models/ExtensionConfig.java (hunk 1 above)
Comment (senthuran16) [reply components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/config/mapping/models/ExtensionConfig.java]: Updated in https://github.com/wso2/carbon-analytics/pull/1840/commits/687d9d726cb4be51cef81b489166d694c78b91dd

File: components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java (hunk 2 above)
Comment (senthuran16) [reply components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java]: Updated in https://github.com/wso2/carbon-analytics/pull/1840/commits/687d9d726cb4be51cef81b489166d694c78b91dd

File: components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java (hunk 3 above)
Comment (senthuran16) [reply components/org.wso2.carbon.siddhi.extensions.installer.core/src/main/java/org/wso2/carbon/siddhi/extensions/installer/core/execution/SiddhiAppExtensionUsageDetectorImpl.java]: Updated in https://github.com/wso2/carbon-analytics/pull/1840/commits/687d9d726cb4be51cef81b489166d694c78b91dd

**Pull request merged**